cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll

```diff
@@ -410,7 +515,40 @@ predicate adjacentDefRead(DefOrUse defOrUse1, UseOrPhi use) {
   )
 }
 
-private predicate useToNode(UseOrPhi use, Node nodeTo) { nodeTo = use.getNode() }
+/**
+ * Holds if `defOrUse` should flow to the final use of the
+ * global variable use represetned by `globalUse`.
+ */
+private predicate defOrUseToGlobalUse(DefOrUse defOrUse, GlobalUse globalUse) {
+  exists(IRBlock bb1, int i1, IRBlock bb2, int i2, SourceVariable v |
+    defOrUse.asDefOrUse().hasIndexInBlock(bb1, i1, v) and
+    globalUse.hasIndexInBlock(bb2, i2, v) and
+    adjacentDefRead(_, pragma[only_bind_into](bb1), pragma[only_bind_into](i1),
+      pragma[only_bind_into](bb2), pragma[only_bind_into](i2))
+  )
+}
+
+/**
+ * Holds if `globalDef` represents the initial definition of a global variable that
+ * flows to `useOrPhi`.
+ */
+private predicate globalDefToUse(GlobalDef globalDef, UseOrPhi useOrPhi) {
+  exists(IRBlock bb1, int i1, IRBlock bb2, int i2, SourceVariable v |
+    globalDef.hasIndexInBlock(bb1, i1, v) and
+    adjacentDefRead(_, pragma[only_bind_into](bb1), pragma[only_bind_into](i1),
+      pragma[only_bind_into](bb2), pragma[only_bind_into](i2)) and
+    useOrPhi.asDefOrUse().hasIndexInBlock(bb2, i2, v)
+  )
+}
+
+private predicate useToNode(UseOrPhi use, Node nodeTo) {
+  exists(OperandBasedUse useImpl |
+    useImpl = use.asDefOrUse() and
+    nodeHasOperand(nodeTo, useImpl.getOperand(), useImpl.getIndirectionIndex())
+  )
+  or
+  nodeTo.(SsaPhiNode).getPhiNode() = use.asPhi()
```

Comment:
I think this is the right way to solve this merge conflict (it's also consistent with how it's done on the current feature [branch](https://github.com/github/codeql/blob/mathiasvp/replace-ast-with-ir-use-usedataflow/cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll#L413)).

```suggestion
  nodeTo = use.getNode()
```

Otherwise, only uses that have a corresponding `Operand` can be mapped to a node. Now, it may actually be that this is semantially identical to the fix you did. This is because the only uses that are not `Operand` based are the uses for global variables at the end of function bodies that this PR is adding, and they don't need to be mapped to a `OperandNode` (but exists so that SSA do not conclude that a definition of global variables is dead).

Reply:
Fixed in 136b5d189c7f23fdd3b473cd4d4eb27612e64332.